### PR TITLE
Refactor StorageHDFS and StorageFile virtual columns filtering

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Filter virtual columns for StorageHDFS and StorageFile from the plan filter condition.
Needed for #53705

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 5521e5d9b16a7527d81cf97742c548570769d143 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58353/5521e5d9b16a7527d81cf97742c548570769d143/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58353/5521e5d9b16a7527d81cf97742c548570769d143/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by Avogar** [comment]
See also my comments in https://github.com/ClickHouse/ClickHouse/pull/56668
Let's refactor all `S3, HDFS, File, URL, Azure` to the new filtering with `ActionsDAG`, and we can remove old filtering by path/file methods from `VirtualColumnUtils` as they are used only in these storages

**Comment by KochetovNicolai** [comment]
Yes, probably I will do it in this PR

File: src/Storages/StorageS3.cpp [hunk 1]
```diff
@@ -1217,19 +1208,34 @@ void StorageS3::read(
...
         throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Reading from a partitioned S3 storage is not implemented yet");
 
-    auto virtual_columns = storage.getVirtuals();
-    auto read_from_format_info = prepareReadingFromFormat(column_names, storage_snapshot, storage.supportsSubsetOfColumns(local_context), virtual_columns);
+    createIterator(nullptr);
 
-    std::shared_ptr<StorageS3Source::IIterator> iterator_wrapper = createFileIterator(
-        query_configuration, storage.distributed_processing, local_context, query_info.query, filter_dags,
-        virtual_columns, nullptr, local_context->getFileProgressCallback());
+    auto read_from_format_info = prepareReadingFromFormat(column_names, storage_snapshot, storage.supportsSubsetOfColumns(local_context), virtual_columns);
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:1238]: `prepareReadingFromFormat` is called twice for no reason (affter https://github.com/ClickHouse/ClickHouse/pull/56668). Once in `StorageS3::read` and second one here. Let's do it once in `StorageS3::read` and pass `read_from_format_info` to `ReadFromStorageS3Step` through its constructor.

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -1165,12 +1156,12 @@ static std::shared_ptr<StorageS3Source::IIterator> createFileIterator(
...
         /// Iterate through disclosed globs and make a source for each file
         return std::make_shared<StorageS3Source::DisclosedGlobIterator>(
-            *configuration.client, configuration.url, query, virtual_columns,
+            *configuration.client, configuration.url, predicate, virtual_columns,
             local_context, read_keys, configuration.request_settings, file_progress_callback);
     }
     else
     {
-        Strings keys = filterKeysForPartitionPruning(configuration.keys, configuration.url.bucket, virtual_columns, filter_dags, local_context);
+        Strings keys = filterKeysForPartitionPruning(configuration.keys, configuration.url.bucket, virtual_columns, predicate, local_context);
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:1164]: I don't think we need functions `filterKeysForPartitionPruning` and `getBlockWithVirtuals` added in https://github.com/ClickHouse/ClickHouse/pull/56668. We can use `VirtualColumnUtils::filterByPathOrFile` in `StorageS3Source::KeysIterator` as it was before

File: src/Storages/StorageS3.cpp [hunk 3]
```diff
@@ -1264,17 +1270,17 @@ void ReadFromStorageS3Step::initializePipeline(QueryPipelineBuilder & pipeline,
...
+    if (pipe.empty())
+        pipe = Pipe(std::make_shared<NullSource>(read_from_format_info.source_header));
 
+    for (const auto & processor : pipe.getProcessors())
+        processors.emplace_back(processor);
 
-void ReadFromStorageS3Step::applyFilters()
-{
-    /// We will use filter_dags in filterKeysForPartitionPruning called from initializePipeline, nothing to do here
+    pipeline.init(std::move(pipe));
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:1283]: Just curious why we need it

Comment (KochetovNicolai) [reply src/Storages/StorageS3.cpp:1283]: Please clarify: need what?
The refactoring is needed for analyzer mainly, but it also makes virtual columns analysis in a proper way (using AST is not valid; you can't just take WHERE part cause e.g. it is applied after JOIN).

`applyFilters` itself is needed to build sets for IN which may be needed; and we have to do it before building pipeline

Comment (Avogar) [reply src/Storages/StorageS3.cpp:1286]: I am about change:
```
pipeline.init(Pipe::unitePipes(std::move(pipes)));
```
to
```
auto pipe = Pipe::unitePipes(std::move(pipes));
for (const auto & processor : pipe.getProcessors())
    processors.emplace_back(processor);
pipeline.init(std::move(pipe));
```
Sorry for confusion

Comment (KochetovNicolai) [reply src/Storages/StorageS3.cpp:1286]: Ah, this one is for proper EXPLAIN PIPELINE
`processors` is a list of processors we create in the source.

**Review by Avogar — APPROVED** [review]

**Pull request merged**